muesli/web/viewsGrading.py

```diff
@@ -201,6 +207,66 @@ def __call__(self):
         self.request.javascript.add('jquery/jquery.fancybox.min.js')
         #grades = {key: value for key,value in grades.items()}
 
+        #create the diagram using pyplot
+
+        #get a list of calculated grades
+        #if no grade is calculated yet then grades[student_id]['calc'] contains an empty string and would cause an exception when converted to a float
+        grades_list = [float(grades[student_id]['calc']) for student_id in grades.keys() if not grades[student_id]['calc'] == '']
+
+        if len(grades_list) > 0:
+
+            #count occurences of grades and save it in a list as tuple (grade, count)
+            tuple_list = list(Counter(grades_list).items())
+
+            #sort the list by grades
+            tuple_list = sorted(tuple_list, key=lambda x: x[0])
+
+            labels = [x[0] for x in tuple_list]
+            values = [x[1] for x in tuple_list]
+
+            indexes = numpy.arange(len(labels))
+            width = 1
+
+            pyplot.rcParams.update({'font.size': 20})
+
+            fig = pyplot.figure(figsize=(12, 9))
+
+            ax = fig.add_subplot(111)
+
+            pyplot.sca(ax)
+            pyplot.bar(indexes, values, width, edgecolor='black', color='red')
+            pyplot.xticks(indexes + width - 1, labels)
+            pyplot.xlabel('Note')
+            pyplot.ylabel('Anzahl')
+
+            yint = range(min(values), math.ceil(max(values)) + 1, math.ceil(max(values)/10))
+            pyplot.yticks(yint)
+
+            output = io.BytesIO()
+            fig.savefig(output, format='png', dpi=50, bbox_inches='tight')
+            pyplot.close(fig)
+            #encode image as base64 so it can be displayed using html
+            encoded_diagram = base64.b64encode(output.getvalue())
+            output.close()
+
+            percentage_message = []
+
+            grades_count = len(grades_list)
+            percentage_list = []
+            for x in range(1, 5):
+                percentage_list.append(100*sum(grade <= x for grade in grades_list)/grades_count)
+
+            percentage_message.append(str(percentage_list[0]) + '% haben die Note 1.0\n')
+            percentage_message.append(str(percentage_list[1]) + '% haben die Note 2.0 oder besser\n')
+            percentage_message.append(str(percentage_list[2]) + '% haben die Note 3.0 oder besser\n')
+            percentage_message.append(str(percentage_list[3]) + '% haben die Note 4.0 oder besser\n')
+            percentage_message.append(str(100-percentage_list[3]) + '% haben die Note 5.0\n')
+
+        else:
+            #empty png image encoded in base64
+            encoded_diagram = 'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8/5+hHgAHggJ/PchI7wAAAABJRU5ErkJggg=='
```

Comment:
I know, this sounds funny, but I am honestly impressed. I had a good laugh xD

Please re-implement this functionality in a new function as a new route. Use the relevant template to load the image from that route. If len(grades_list) == 0, simply return HTTP status code 404, else return the image in binary form. This avoids using base64 encoding.

Reply:
> If len(grades_list) == 0
```
if not grades_list:
```

Reply:
This was done here to avoid re-querying the database and re-calculating the grades an additional time every time the formula for the grades was changed.

Reply:
ah, okay. I can see, why you would like to do that. I don't think the additional database query is a performance issue here. The problem is, that this is very unreadable, breaks consistency, uses a base64 blob, etc.
Please implement this in the way I described above.